### PR TITLE
ScScore Porting

### PR DESCRIPTION
**Pull request by aaronrockmenezes**
## Description

ScScore Porting to PyTorch

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/scscore.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    The learned metric (SCScore) exhibits highly desirable nonlinear behavior, particularly in recognizing increases in synthetic complexity throughout a number of linear synthetic routes.
+    The SCScore model can accurately predict the synthetic complexity of a variety of molecules, including both drug-like and natural product molecules.
+    SCScore has the potential to be a valuable tool for chemists who are working on drug discovery and other areas of chemistry.
+
+    The learned metric (SCScore) exhibits highly desirable nonlinear behavior, particularly in recognizing increases in synthetic complexity throughout a number of linear synthetic routes.
+
+    Our model uses hingeloss instead of the shifted relu loss as in the supplementary material [2]_ provided by the author.
+    This could cause differentiation issues with compounds that are "close" to each other in "complexity".
+
+    The default values for the model are the same as the ones used in the original paper [1]_.
```
Comment (rbharath) [inline deepchem/models/torch_models/scscore.py:112]: Usage example

File: deepchem/models/torch_models/scscore.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                        dropout=dropout, 
+                        score_scale=score_scale)
+        
+        loss = HingeLoss()
+        
+        super(ScScoreModel, self).__init__(model, 
+                                           loss=loss, 
+                                           **kwargs)
+    
+    def default_generator(self,
```
Comment (rbharath) [inline deepchem/models/torch_models/scscore.py:157]: You can just remove this, inheritance will work out

Comment (aaronrockmenezes) [reply deepchem/models/torch_models/scscore.py:157]: Will do

File: deepchem/models/torch_models/scscore.py (hunk 1 above)
Comment (aaronrockmenezes) [reply deepchem/models/torch_models/scscore.py:106]: Done

**Review by rbharath — APPROVED** [review]
LGTM

@aaronrockmenezes Can you remove the draft marking and verify that no related tests are failing?

**Comment by aaronrockmenezes** [comment]
@rbharath I want to discuss another feature for this model. Can we discuss that at OH?
We discussed earlier that the model itself isn't too useful. I found the trained weights for the model, so I was wondering if we could keep a method to maybe pretrain the model?

**Comment by aaronrockmenezes** [comment]
@rbharath For converting the model to ModularTorchModel, we need to break the layers up into components. Do you want me to club the hidden layers as 1 component while keeping the input and output layers as 2 more components? Or should I keep all the layers as individual components?

File: deepchem/models/torch_models/scscore.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        if score_scale < 1:
+            raise ValueError(
+                f"Score scale must be greater than 1, but got {score_scale}")
+
+        model = ScScore(n_features=n_features,
+                        layer_sizes=layer_sizes,
+                        dropout=dropout,
+                        score_scale=score_scale)
+
+        loss = HingeLoss()
```
[the remark this answers is not in the document]
Comment (shreyasvinaya) [reply deepchem/models/torch_models/scscore.py:163]: I would suggest you to allow the user to edit it and keep hinge as the default

**Review by rbharath — COMMENTED** [review]
To summarize, may be easier to just use TorchModel for now. ModularTorchModel probably needs some thought for non-self-supervised models to be friendly

File: deepchem/models/torch_models/scscore.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
+from deepchem.models.torch_models.modular import ModularTorchModel
+from deepchem.models.losses import HingeLoss
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class ScScore(nn.Module):
```
[the remark this answers is not in the document]
Comment (aaronrockmenezes) [reply deepchem/models/torch_models/scscore.py:7]: Added. Can you review? @shreyasvinaya

**Comment by aaronrockmenezes** [comment]
@rbharath Made the Hinge Loss default if user doesn't give loss function

**Comment by shreyasvinaya** [comment]
@aaronrockmenezes The doctest seems to be failing could you please fix it
```
FAILED deepchem/models/torch_models/scscore.py::deepchem.models.torch_models.scscore.ScScore
FAILED deepchem/models/torch_models/scscore.py::deepchem.models.torch_models.scscore.ScScoreModel
```

**Comment by aaronrockmenezes** [comment]
@shreyasvinaya I've fixed the doctest failures.

File: deepchem/models/torch_models/scscore.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
+from deepchem.models.torch_models import TorchModel
+from deepchem.models.losses import HingeLoss
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class ScScore(nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/scscore.py:7]: In a follow up PR, please add the new model to the docs

**Review by rbharath — APPROVED** [review]
LGTM
